oremda/pipeline/runner/context.py

```diff
@@ -0,0 +1,72 @@
+from contextlib import contextmanager
+
+from typing import Dict, Set
+
+from pydantic import BaseModel, Field
+
+from oremda.typing import IdType
+from oremda.clients import Client as ContainerClientFactory
+from oremda.clients.base import ClientBase as ContainerClient
+from oremda.plasma_client import PlasmaClient
+from oremda.registry import Registry
+from oremda.constants import DEFAULT_OREMDA_VAR_DIR
+from oremda.utils.plasma import start_plasma_store
+from oremda.typing import ContainerType
+from oremda.pipeline.runner.config import settings
+
+from oremda.models import (
+    SessionModel,
+    WebsocketModel,
+)
+
+from oremda.pipeline.models import (
+    PipelineModel,
+)
+
+
+class GlobalContext(BaseModel):
+    plasma_client: PlasmaClient = Field(...)
+    container_client: ContainerClient = Field(...)
+    registry: Registry = Field(...)
+    sessions: Dict[IdType, SessionModel] = {}
+    pipelines: Dict[IdType, PipelineModel] = {}
+    websockets: Dict[IdType, WebsocketModel] = {}
+    session_pipelines: Dict[IdType, Set[IdType]] = {}
+    session_websockets: Dict[IdType, Set[IdType]] = {}
+
+    class Config:
+        arbitrary_types_allowed = True
```

Comment:
Nice! I like having a global context. This can seriously reduce the number of arguments we have to pass around the program.

Maybe we should eventually start using this in some other places where we are passing around arguments?

Reply:
The global context here serves the purpose of replacing a future database and it works fine.

But in general I would be strongly against replacing regular function parameters for some global state.

Reply:
We have global states in ParaView, Tomviz, Hexrd, etc. They can be very effective if done well. If there is good reason to believe a variable should only have one instance throughout the whole program, then moving it into a global state (such as singleton) can rapidly speed up R&D and the amount of work you can accomplish. The costs of maintaining arguments throughout the program that should be global can be high.